### DISCUSSION TITLE
xUnit v3 and some stuff about TUnit

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#126] [General]
# xUnit v3 and some stuff about TUnit

I know that there is a lot of fuse about TUnit and I am here writing about xUnit. I might cover TUnit in the future, but for now, the topic is the v3 prerelease of xUnit!

https://steven-giesel.com/blogPost/669d5aa2-b86a-4096-948c-9a767b7071e3/xunit-v3-and-some-stuff-about-tunit

**Comment by tabrizgulmammadov** [comment]
Thanks for your articles. It's my routine to read your articles each day as my first job. Interestingly, what happened, is that you used our (Azerbaijan alphabet) 'ü' character in the word "disüose"))

**Reply by linkdotnet** [reply]
Hey hey - thank you for the kind words! Really means a lot. 
I fixed the error. 

Interesting though - I am German and using a Germany layout where the „ü“ is next to „p“, this is how it happens :)
I hope disüose doesn’t mean anything :):):)

**Comment by toupswork** [comment]
On the topic of testing, I wanted to get your thoughts on this. I found that my team spends a lot of time writing unit tests. The pain points are that all of the mocking libraries can't mock sealed or static methods (specifically extension methods), and even for the ones that can be mocked, we end up writing a ton of mocks and fake instances of models, many with huge ctors and are deeply nested. I figured out an idea that has saved us so much time. I decided to write a mock class that can clone, instantiate an instance of any object by bypassing the ctor (then you can set only properties used in the test) and It can redirect one method, regardless of what it is to another method. It cuts down on the time it takes to write unit tests by about 80%. It's super cool. I can show you the code if you're interested, but basically it goes to the method table of a type, gets the method address, overwrites the preamble of the method with x64 assembly that loads the destination address into the RCX register and then JMP to RCX. I do this rather than overwriting the pointer because patching the preamble prevents tiered PGO from recompiling the source method.

**Reply by linkdotnet** [reply]
Wow - that is advanced :)

so if I understand you correctly, if I would run those tests on my Mac, they wouldn’t pass as it doesn’t use x64 as arch, correct?

For sure it is one way. I hope that someone can maintain and understand it while you are on vacation.

Maybe another question: Why are those objects so deeply nested? And why is it a problem? For example, if you test one laser above all that, you basically have less overhead (API Testing for example).

**Reply by toupswork** [reply]
Well our servers are x64, so if the day comes where we run on Apple Silicone, then I'll add an ARM 64 stub. 
It has remained stable since the introduction of .NET Core. 

Why deeply nested? Well some of our domain models are. For example, a Contact model has phones numbers, emails, addresses, permissions, and so on. Having to create fake instances would require a lot of arguments passed to the ctor when only a handful are used by the test.

Take a look https://github.com/toupswork/MockHelpers/blob/main/MockHelper.cs

**Reply by linkdotnet** [reply]
Super interesting and way beyond my horizon of things I did. Nevertheless interesting.

> so if the day comes where we run on Apple Silicone, then I'll add an ARM 64 stub.

The more probable way is that a dev has an M-Chip for example.

> Why deeply nested? Well some of our domain models are. For example, a Contact model has phones numbers, emails, addresses, permissions, and so on. 

Depending on the context, I try to resort to some way of a fluent builder. So `new MyDomainEntityBuilder().Build();` would already produce a valid domain object with sensible default values. Until now that worked quite well in many projects (for me).
